src/java.base/share/classes/jdk/internal/ref/PhantomCleanable.java

```diff
@@ -44,14 +44,21 @@ public abstract class PhantomCleanable<T> extends PhantomReference<T>
         implements Cleaner.Cleanable {
 
     /**
-     * Links to previous and next in a doubly-linked list.
+     * The list of PhantomCleanable; synchronizes insert and remove.
      */
-    PhantomCleanable<?> prev = this, next = this;
+    private final CleanerImpl.CleanableList list;
 
     /**
-     * The list of PhantomCleanable; synchronizes insert and remove.
+     * Node for this PhantomCleanable in the list.
+     * Synchronized by the same lock as the list itself.
      */
-    private final PhantomCleanable<?> list;
+    CleanerImpl.CleanableList.Node node;
```

Comment:
I think we can get away with not storing a `Node` in every `PhantomCleanable`, and instead look for it at `index` in each `Node` in the list. But I don't know if this would be a performance win. Heap savings (in `PhantomCleanableRef`)? Worse locality from loading multiple `Nodes`?
Just an idea...

Reply:
I don't think this is a great idea: it means linear search under the lock, which I don't think we want. As it stands now, we are doing removals in constant time. Current version also affects the footprint positively already: we change 3 references into 2 references + 1 int, so there is no pressing need for smartness in optimizing the footprint.